File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        def postorder(node, maxDepth):
            if not node:
                return 0
            
            left = postorder(node.left, maxDepth)
            right = postorder(node.right, maxDepth)
            maxDepth = max(maxDepth, max(left, right) + 1)
            
            return max(left, right) + 1
        
        maxDepth = 0
        return postorder(root, maxDepth)

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        int maxDepth = 0;
        maxDepth = postorder(root, 0);
        return maxDepth;
    }

    public int postorder(TreeNode node, int maxDepth) {
        if(node == null) {
            return 0;
        }

        int left = postorder(node.left, maxDepth);
        int right = postorder(node.right, maxDepth);
        maxDepth = Math.max(maxDepth, Math.max(left, right) + 1);

        return Math.max(left, right) + 1;
    }
}